docs/new-editor.md

```diff
@@ -0,0 +1,163 @@
+[`workspace/didChangeConfiguration`](https://microsoft.github.io/language-server-protocol/specification#workspace_didChangeWatchedFiles)
+notifications to allow end-users to control behavior of the server. The editor
+client is expected to send a `workspace/didChangeConfiguration` notification
+containing user configuration right after the `initialized` notification.
+
+A full list of supported configuration options can be found in the VS Code
+plugin
+[package.json](https://github.com/scalameta/metals/blob/master/vscode-extension/package.json)
+manifest. Some configuration options are required by the server while others
+should be handled by the client.
+
+Server side configuration options include settings to enable
+experimental/unstable features such as completions with the presentation
+compiler. Observe that even if metals registers the capability for completions
+during the `initialize` handshake, the end-user must opt-into enabling
+completions with the configuration setting `scalac.completions.enabled=true`. If
+completions are disabled in the configuration
+(`scalac.completions.enabled=false`, default value), then metals responds with
+an empty list of completion suggestions.
+
+Client side configuration options include
+
+* `serverVersion: String`: while metals is still under active development, it is
+  recommended to allow end-users to easily configure the version of the metals
+  server.
+* `scalafmt.onSave: Boolean`: if true, runs scalafmt on
+  [`textDocument/willSaveWaitUntil`][] client requests.
+
+## Commands
+
+Metals uses [`workspace/executeCommand`][] requests to allow end-users to
+trigger actions in the server on-demand. The full list of supported commands by
+the server can be found in [WorkspaceCommand.scala][]. See the VS Code plugin
+[package.json][] manifest how these commands are configured in VS Code. Observe
+that `restartServer` must be handled client-side.
+
+## Code actions
+
+The metals server uses [`textDocument/codeAction`][] requests from the editor
+client to provide passive refactoring hints with scalafix. Currently, only the
+"removed unused import" refactoring is supported but more refactorings may be
+added in the future.
+
+<img src="img/code-actions.png" align="right" width="150px" style="padding-left: 20px"/>
+
+The sequence diagram for refactoring hints is quite involved.
+
+1.  Server publishes warning diagnostics about "Unused import"
+2.  Client sends a code action request with the "Unused import" warning as part
+    of `context.diagnostics`. Server responds with instructions to execute the
+    `scalafixUnusedImport` command.
+3.  User triggers the `scalafixUnusedImports` command on-demand.
+4.  Server sends request to apply a set of text edits. Client responds if edit
+    was successfully applied or not.
+5.  Server responds to `workspace/executeCommand` request.
+
+In VS Code, code actions are suggested to the user via light bulbs when hovering
+above the "Unused import" warnings.
```

Comment:
...and can be triggered with the <kbd>CMD</kbd> + <kbd>.</kbd> shortcut (on macOS)